src/read.c

```diff
@@ -211,185 +211,444 @@ static Obj GAPInfo;
+            break;
+        case R_ELM_COMOBJ_EXPR:
+            IntrElmComObjExpr();
+            break;
+        case R_FUNCCALL:
+            IntrFuncCallEnd(needExpr, 0, ref.narg);
+            break;
+        case R_FUNCCALL_OPTS:
+            IntrFuncCallEnd(needExpr, 1, ref.narg);
+            break;
+        default:
+            // This should never be reached
+            SyntaxError("Parse error in EvalRef");
         }
+    }
+    return 0;
+}
 
-        /* <Var> '{' <Expr> '}'  sublist selector                          */
-        else if ( STATE(Symbol) == S_LBRACE ) {
-            Match( S_LBRACE, "{", follow );
-            ReadExpr( S_RBRACE|follow, 'r' );
-            Match( S_RBRACE, "}", follow );
-            type = (level == 0 ? '{' : '}');
+void AssignRef(const LHSRef ref)
+{
+    TRY_READ
+    {
+        switch ((UInt)ref.type) {
+        case R_LVAR:
+            IntrAssLVar(ref.var);
+            break;
+        case R_HVAR:
+            IntrAssHVar(ref.var);
+            break;
+        case R_DVAR:
+            IntrAssDVar(ref.var, ref.nest0);
+            break;
+        case R_GVAR:
+            IntrAssGVar(ref.var);
+            break;
+        case R_ELM_LIST:
+            if (ref.level == 0)
+                IntrAssList(ref.narg);
+            else
+                IntrAssListLevel(ref.narg, ref.level);
+            break;
+        case R_ELMS_LIST:
+            if (ref.level == 0)
+                IntrAsssList();
+            else
+                IntrAsssListLevel(ref.level);
+            break;
+        case R_ELM_POSOBJ:
+            if (ref.level == 0)
+                IntrAssPosObj();
+            else
+                IntrAssPosObjLevel(ref.level);
+            break;
+        case R_ELMS_POSOBJ:
+            if (ref.level == 0)
+                IntrAsssPosObj();
+            else
+                IntrAsssPosObjLevel(ref.level);
+            break;
+        case R_ELM_REC_NAME:
+            IntrAssRecName(ref.rnam);
+            break;
+        case R_ELM_REC_EXPR:
+            IntrAssRecExpr();
+            break;
+        case R_ELM_COMOBJ_NAME:
+            IntrAssComObjName(ref.rnam);
+            break;
+        case R_ELM_COMOBJ_EXPR:
+            IntrAssComObjExpr();
+            break;
+        case R_FUNCCALL:
+            IntrFuncCallEnd(0, 0, ref.narg);
+            break;
+        case R_FUNCCALL_OPTS:
+            IntrFuncCallEnd(0, 1, ref.narg);
+            break;
+        default:
+            // This should never be reached
+            SyntaxError("Parse error in AssignRef");
         }
+    }
+}
 
-        /* <Var> '![' <Expr> ']'  list selector                            */
-        else if ( STATE(Symbol) == S_BLBRACK ) {
-            Match( S_BLBRACK, "![", follow );
-            ReadExpr( S_RBRACK|follow, 'r' );
-            Match( S_RBRACK, "]", follow );
-            type = (level == 0 ? '<' : '>');
+void UnbindRef(const LHSRef ref)
+{
+    volatile enum REFTYPE type = ref.type;
```

Comment:
For the same reason the code this replaces uses `volatile`: because the `TRY_READ` macro uses `setjmp`, and gcc on Linux then warns for non-volatile local variables that they might be clobbered by that.